### **STAT 220 Probability Lab**

# Function A

A. Sum Greater Function: Calculate the probability that the sum of m dice (each with n sides) is greater than x.

In [1]:
# m = number of dice (3)
# n = number of sides on each dice (6)
# x = sum (14)

import itertools
def prob_sum_greater(m, n, x):
    rolls = itertools.product(range(1, n+1), repeat=m) #list all possible rolls
    total = 0
    favorable = 0
    for roll in rolls:
        total += 1  # increment total for each possible roll
        if sum(roll) > x:
            favorable += 1
    return favorable / total  # probability = favorable outcomes / total outcomes

Example: For 3d6, what is the probability that the sum is at least 14?

In [2]:
# To use the pro_sum_greater function we need to know m(the number of dice), n(number of sides on each dice), and x(the sum).

# Lab defines this function as > but scenarios use ≥
# Since dice sums are integers, ≥ x is the same as > (x-1), so we put in x-1 for x (example: ≥ 14 --> > 13)
prob_a = prob_sum_greater(3, 6, 13)

# we multiply by 100 to get our percent
print(f"{prob_a*100:.2f}%")

16.20%


# Function B

B. At Least One Function: Calculate the probability that, when rolling m dice (each with n sides), at least one die shows a value greater than or equal to x.

In [3]:
# using the complement rule

def prob_at_least_one(m, n, x):
    return 1 - ((x - 1) / n) ** m

Example: For 4d8, what is the probability at least one die shows an 8?

In [4]:
m = 4
n = 8
x = 8

prob = prob_at_least_one(m, n, x)
print(f"Probability of at least one roll being >= {x} is: {prob:.4f}")

Probability of at least one roll being >= 8 is: 0.4138


Explanation:

We used the complement rule for this function. The chance one die shows a value less than x is (x − 1) / n. Because the (m) dice are independent, the probability that all the dice show less than x is ((x − 1) / n)^m. Thus, the probability that at least one die is ≥ x is 1 - ((x − 1) / n)^m. In the example 4d8 where at least one die shows an 8 or greater (x = 8), that means 1 − (7/8)^4 ≈ 0.4138.

# Function C

C.
Exactly r Maximums Function: Calculate the probability that, when rolling m dice (each
with n sides), exactly r of them equal the maximum possible value n.

In [5]:
import math

def prob_exactly_r_max(m, n, r):
  return math.comb(m, r) * (1/n)**r * ((n-1)/n)**(m-r)


(m, r) chooses which dice show the maximum.

(1/n)^r is the probability those r dice show maximum.

(n-1/n)^(m-r) is the probability the remaining dice don't show maximum.

Example: For 3d6, what
is the probability that exactly two dice show a 6?

In [6]:
example = prob_exactly_r_max(3, 6, 2)
print(f"Probability that exactly two dice show a 6: {example:.4f} ({example*100:.2f}%)")

Probability that exactly two dice show a 6: 0.0694 (6.94%)


# Function D

D. Tries Until Threshold Function: Model the number of independent trials until rolling at least x on a single 1dn die.

In [7]:
def prob_until_threshold(n, x):
  #x = target roll
  #n = number of sides on die

    if (n < 1 or x < 1): #not valid die
        print("n and x must be >= 1")
        return None

    if (x > n): #result can never be larger than number of sides on die
        print(f"never able to roll {x} or higher on a {n}dn die")
        return None

    p = (n-x+1)/n #using geometric distribution
    expectedRolls = 1/p #expected number of rolls until first success = 1/p
    return expectedRolls

Example: For a 1d20, what is the expected number of rolls until you get 17 or higher?

In [8]:
print(f"Expected number of rolls = {prob_until_threshold(20, 17)}")

Expected number of rolls = 5.0


Explanation for prob_until_threshold(n, x):

Each roll is independent. For this, success means rolling ≥ x on a 1dn die. On a roll, the chance of success is p = (n − x + 1) / n because there are n − x + 1 success sides (x through n) out of n total sides on the die. The number of rolls until the first success uses the geometric distirbution, so the expected number of rolls is 1 / p. From the example 1d20 with x = 17, p = 4/20 = 0.20 and so the expected number of rolls = 1/0.20 = 5.


# Evaluation Scenario

• Roll 1d20 and get a result of 17 or more

In [9]:
#1d20 ≥ 17, so use > 16 (x = 16)
p_1 = prob_sum_greater(1, 20, 16)
print(f"{p_1*100:.2f}%")

20.00%


• Roll 3d4 and get a result of 10 or more

In [10]:
#3d4 ≥ 10, so use use > 9 (x = 9)
p_2 = (prob_sum_greater(3, 4, 9))
print(f"{p_2*100:.2f}%")

15.62%


• Roll 4d6 and have at least one die show a 6

In [11]:
p_3 = (prob_at_least_one(4, 6, 6))
print(f"{p_3*100:.2f}%")

51.77%


# Which Scenario Has the Best Odds?

For this lab evaluation, we compared three scenarios using our functions (A & B):

1. 1d20 ≥ 17 = 20.00%

2. 3d4 ≥ 10 = 15.62%

3. 4d6 (at least one 6) = 51.77%

The best option is rolling 4d6 and getting at least one die to show a 6 or greater because it has the highest probability (51.77%). This makes sense because with four dice there are more chances to roll a 6. This probability function uses the complement rule, so the chance of no sixes is (5/6)^4, and thus the success chance is 1−(5/6)^4 ≈ 51.77%. The other two scenarios have to hit a higher threshold using less dice, which is ultimately less likely.